### PR TITLE
Support broadcast signal for signal end events

### LINKED ISSUE #11920: Support Broadcast signal for Signal End Events
- ## Adjust `EndEventProcessor` to broadcast a signal on activation
	- Introduce a new `SignalEndEventBehavior`
	- When Signal End Event activates:
		- Apply input mappings
		- Transition to activated
		- Write `Signal:Broadcast` command
		- Apply output mappings
		- Transition to complete the element

### PR DESCRIPTION
**Pull request by skayliu**
## Description

- ## Adjust `EndEventProcessor` to broadcast a signal on activation
	- Introduce a new `SignalEndEventBehavior`
	- When Signal End Event activates:
		- Apply input mappings
		- Transition to activated
		- Write `Signal:Broadcast` command
		- Apply output mappings
		- Transition to complete the element

## Related issues

closes #11920 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [x] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [x] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [x] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by skayliu** [comment]
Hey, @korthout. Please take a moment to check this out, When you have time,. 🙇

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java:69]: Calling the constructor of new records can be costly and should be avoided during regular processing. Instead, we create such records as fields only to be constructed once. We can reuse it where we need it using the reset/wrap methods.

❌ Please reuse a single record instead of constructing a new one every time

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java:61]: 🔧 I would prefer to flip the order (first try to use the expression, if it's not available then use the static name).

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnBehaviors.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnBehaviors.java:42]: 👍 Great idea to add a signal behavior that can be reused for signal intermediate throw events.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEndEvent.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEndEvent.java:74]: 👍 Nice catch!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEndEvent.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/ExecutableEndEvent.java:88]: 🔧 We can move this method closer to the other `isXEndEvent()` methods. That keeps easy to read.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/signal/SignalEndEventTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/signal/SignalEndEventTest.java:78]: ❌ Let's ensure we only look for the `BROADCASTED` event. 

🔧 The name does not matter in this look-up, as long as we assert it below.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java:72]: ❌ Instead of writing an event to the state, we'll need to write a command. Please use the command writer instead.

💭 If we don't write a command, then it becomes harder to distribute the signal broadcast later.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnSignalBehavior.java:49]: 💭 Users will likely expect the broadcasted signal to contain the variables visible from the element's scope. We can do this in a later pull request.

**Review by korthout — CHANGES_REQUESTED** [review]
Well done @skayliu 👏 

I have a few change requests. Please have a look 🙇

**Comment by skayliu** [comment]
Hey @korthout, I've applyed all you reviews, please check again, Thank you.

**Review by korthout — APPROVED** [review]
Thanks so much, @skayliu! This feature looks awesome! 🚀 

👍 I saw that you also already implemented the ability to pass local variables from the end event to the broadcasted signal. I've asked the team to provide feedback on this separately, as we want to understand better whether this is our users' expected/wanted behavior. I think it's great to have it in, so we can run some manual tests with it.

👍 LGTM

🔧 I have one last suggestion. Feel free to apply it or merge the pull request as it is.

**Review by korthout — APPROVED** [review]
😍 Awesome!

LGTM 👏 

bors merge

---

@skayliu Once this is merged, we should inform the following teams about this new feature (coming in `8.3.0-alpha1`):
- [x] [Operate](https://github.com/camunda/operate/issues) needs to know that a new symbol is available which they must be able to render in their UI
- [x] [Web Modeler](https://github.com/camunda/web-modeler/issues) needs to know that Zeebe supports a new symbol that can be modeled (update the linting errors)
- [x] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues) needs to know that Zeebe supports a new symbol that can be modeled (update the linting errors)
- [x] [Optimize](https://github.com/camunda/camunda-optimize/issues) needs to know that a new symbol is available which they must be able to render in their UI
- [x] [Docs](https://github.com/camunda/camunda-platform-docs/issues) we need an issue there to coordinate that we write some documentation about this new symbol

**Comment by skayliu** [comment]
Hey, @korthout, Do i need to rebase to main and unblock then `signal end event` of [Reject unsupported signal events #12117](https://github.com/camunda/zeebe/pull/12117)

**Comment by korthout** [comment]
> Hey, @korthout, Do i need to rebase to main and unblock then `signal end event` of [Reject unsupported signal events #12117](https://github.com/camunda/zeebe/pull/12117) and delete the `UnsupportedSignalEndEventValidator` class?

Yes, that's a good point! Thanks, I forgot about it :D

**Review by korthout — APPROVED** [review]
Thanks for the last changes!

LGTM 👍 

bors merge

**Pull request merged**
